propagators/aws/xray/Propagator.php

```diff
@@ -0,0 +1,151 @@
+        $randomHex = substr($traceId, self::TRACE_ID_TIMESTAMP_LENGTH);
+        $samplingFlag = $context->isSampled() ? self::IS_SAMPLED : self::NOT_SAMPLED;
+
+        $traceHeader = 'Root=' . self::VERSION_NUMBER . '-' . $timestamp . "-" . $randomHex . ";Parent=" . $spanId. ";Sampled=" . $samplingFlag;
+        $setter->set($carrier, self::AWSXRAY_TRACE_ID_HEADER, $traceHeader);
+    }
+
+    /**
+     * Extracts the span context from the carrier to transfer to the next service
+     * This function will parse all parts of the header and validate that it is
+     * formatted properly to AWS X-ray standards
+     */
+    public static function extract($carrier, API\PropagationGetter $getter): API\SpanContext
+    {
+        //Extract the traceHeader using the getter from the carrier
+        $traceHeader = $getter->get($carrier, self::AWSXRAY_TRACE_ID_HEADER);
+        
+        //If it is null it creates a dummy header with INVALID_TRACE and INVALID_SPAN
+        // restore(string $traceId, string $spanId, bool $sampled = false, bool $isRemote = false, ?API\TraceState $traceState = null): SpanContext
+        if ($traceHeader === null || gettype($traceHeader) !== 'string') {
+            return SpanContext::getInvalid();
+        }
+
+        $headerComponents = explode(';', $traceHeader);
+        $parsedTraceId = SpanContext::INVALID_TRACE;
+        $parsedSpanId = SpanContext::INVALID_SPAN;
+        $isSampled = false;
+
+        foreach($headerComponents as $component) {
+            //the key value pair of a component of the traceHeader. 
+            //Example: key = 'Parent' value = '53995c3f42cd8ad8'
+            $componentPair = explode('=', $component);
+
+            if ($componentPair[0] === 'Parent' && self::isValidSpanId($componentPair[1])) {
+                $parsedSpanId = $componentPair[1];
+            }
+            else if ($componentPair[0] === 'Root') {
+                $traceIdParts = explode('-', $componentPair[1]);
+                if (count($traceIdParts) > 1 && self::isValidTraceId($traceIdParts[1] . $traceIdParts[2]) && $traceIdParts[0] === self::VERSION_NUMBER) {
+                    $parsedTraceId = $traceIdParts[1] . $traceIdParts[2];
+                }
+            }
+            else if ($componentPair[0] === 'Sampled' && $componentPair[1] === '1') {
+                $isSampled = true;
+            }
+        }
+
+        //what is isRemote parameter (its a bool)? should I set it to true or false?
+        return SpanContext::restore($parsedTraceId, $parsedSpanId, $isSampled, true);
+    }
+
+    /**
+     * Checks whether the traceId is valid or not
+     * Length of traceId must be 35 characters: 1+-+8hex+-+24hex
+     * Checks version number, traceId timestamp and random 24 hex string
+     */
+    private static function isValidTraceId($traceId): bool {
+        return preg_match(SpanContext::VALID_TRACE,  $traceId) &&  $traceId !== SpanContext::INVALID_TRACE;
```

Comment:
```suggestion
        return ctype_xdigit($traceId) && strlen($traceId) === 32 &&  $traceId !== SpanContext::INVALID_TRACE;
```

regexes are expensive, this should be a cheaper equivalent.  Do the same for spanId as well.

Reply:
Fixed